File: boj/1260-DFS-BFS-Basic/code.js
// 답은 이게 맞는데...
// BOJ에서 입력 처리를 어떻게 해야할지 모르겠다
const fs = require('fs');
const inp = fs.readFileSync('/dev/stdin').toString().split('\n');

const tree = {};

let n = 0;
let m = 0;
let v = 0;

inp.map((v, i) => {
    const line = v.split(' ');
    if(i != 0) {
        const from = parseInt(line[0]);
        const to = parseInt(line[1]);

        if(tree[from] == undefined) {
            tree[from] = [];
        }
        tree[from].push(to);

        if(tree[to] == undefined) {
            tree[to] = [];
        }
        tree[to].push(from);
    }
    else {
        n = line[0];
        m = line[1];
        v = line[2];
    }
});

// tree 소트 수행
Object.keys(tree).map(v => {
    tree[v].sort();
});

const printVisits = (visits) => {
    let print = "";
    visits.map((v, i) => {
        print += v;
        if(i < visits.length - 1) {
            print += ' ';
        }
    });
    console.log(print);
}

const dfsvisit = [];
DFS = (node) => {
    dfsvisit.push(node);
    const nextnodes = tree[node];
    
    if(nextnodes != undefined) {
        nextnodes.map(v => {
            if(!dfsvisit.includes(v)) {
                DFS(v);
            }
        });
    }
}
DFS(v);
printVisits(dfsvisit);

const bfsvisit = [];
const bfsqueue = [];
BFS = (node) => {
    bfsqueue.push(node);

    while(bfsqueue.length > 0) {
        const next = bfsqueue.shift();

        if(bfsvisit.includes(next)) {
            continue;
        }
        bfsvisit.push(next);

        const nextnodes = tree[next];
        // 이미 방문한 노드가 아니면 queue에 다 집어넣음
        if(nextnodes != undefined) {
            nextnodes.map(v => {
                if(!bfsvisit.includes(v)) {
                    bfsqueue.push(v);
                }
            });
        }
    }
}
BFS(v);
printVisits(bfsvisit);

File: boj/1260-DFS-BFS-Basic/code.kt
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

val tree = HashMap<Int, SortedSet<Int>>()
var dfsvisit = ArrayList<Int>()
var bfsvisit = ArrayList<Int>()
var bfsqueue = LinkedList<Int>()

fun main() {
    val nmv = readLine()!!.split(' ').map(String::toInt)

    val n = nmv[0]
    val m = nmv[1]
    val v = nmv[2]

    for(i in 0 until m) {
        val fromTo = readLine()!!.split(' ').map(String::toInt)

        val from = fromTo[0]
        val to = fromTo[1]

        if(tree.containsKey(from)) {
            tree[from]!!.add(to)
        }
        else {
            tree[from] = TreeSet<Int>()
            tree[from]!!.add(to)
        }

        if(tree.containsKey(to)) {
            tree[to]!!.add(from)
        }
        else {
            tree[to] = TreeSet<Int>()
            tree[to]!!.add(from)
        }
    }

    DFS(v)
    printVisits(dfsvisit)

    BFS(v)
    printVisits(bfsvisit)
}

fun printVisits(visits: ArrayList<Int>) {
    var print = ""
    for(x in visits.indices) {
        print += visits[x].toString()
        if(x < visits.size-1) print += " "
    }
    println(print);
}

fun DFS(node: Int) {
    dfsvisit.add(node);

    if(tree.containsKey(node)) {
        val nextnodes = tree[node]
        for(x in nextnodes!!) {
            if(!dfsvisit.contains(x)) {
                DFS(x)
            }
        }
    }
}

fun BFS(node: Int) {
    bfsqueue.add(node);

    while(bfsqueue.size > 0) {
        val next = bfsqueue.poll()

        if(bfsvisit.contains(next)) {
            continue
        }
        bfsvisit.add(next)

        if(tree.containsKey(next)) {
            val nextnodes = tree[next]!!
            // 이미 방문한 노드가 아니면 queue에 다 집어넣음
            for(x in nextnodes) {
                if(!bfsvisit.contains(x)) {
                    bfsqueue.add(x)
                }
            }
        }
    }
}